### PR TITLE
Experience Issues - Hide non template first themes behind a 'more themes' button.

### LINKED ISSUE #37285: Themes: Place non template-first themes behind a “more themes” option
Switching from a template-first theme to an older theme that does not use Gutenberg templates is extremely painful. It retains the previous template-first theme's page template so the theme does not look anything like the theme they have just activated.

See the "Changing Designs" testing video on this post: p58i-8d4-p2

We must discourage our users from switching to older themes by placing them behind a "More Themes" button in the theme showcase. Only show them when the user hits this button, and include a notice that says "These themes are older, and may require more work to set up.".

### PR DESCRIPTION
**Pull request by Addison-Stavlo**
#### Changes proposed in this Pull Request

* Hide All Themes Search behind a 'More Themes' Button.
* Add a default list of recommended themes (template first) at the top of the Theme selection page.
* Adds a recommendedThemes section to the Themes store and corresponding action for retrieving recommended themes.

Note:  This PR **DOES NOT** introduce all of the style changes based on design mockups for the new themes section.  This address the main blocking issue of hiding old themes behind a button while introducing a list of recommended ones at the top.  Some of these design changes may require some significant restructuring of some of the components functionalities which would be best addressed in follow up PRs.

#### Testing instructions

* Switch to this branch and run `npm start` in wp-calypso.
* Once built, navigate to calypso.localhost.
* Navigate to the Themes section and ensure the list of recommended Themes renders at the top of the page.
* Verify you can access and switch to these themes.
* Scroll down and click the "Show All Themes" button.
* Verify that the search functionality and selecting themes works as intended.
* Do this for both simple and atomic sites.

**Things to Consider**
Since there are so many special cases that can contribute to different components showing and not showing, its a bit hard to tell if everything is in the correct place.  If we can think of more special cases that would cause the Themes section to display differently, please note them and we can test those as well.
* Are we hiding any components behind this button that we do not want hidden? (Query Theme Filters, Upwork Banner, Upsell Banner, etc.)
* Are we showing any components that should be hidden?  (Page View Tracker, Sub Masterbar Nav, etc.)
* Are there any other things we need to test or consider for this?  This is a section of the codebase that has a lot of "moving pieces" and specialized functionality while the impact of this change would be live to all calypso users.

**Before**
Screen Shot 2019-11-21 at 7 34 13 PM

**After - Top Of Page**
Screen Shot 2019-11-21 at 7 16 12 PM

**After - After Clicking Button**
Screen Shot 2019-11-21 at 7 16 35 PM

Note that the above screenshot is on an Atomic site and that the 'Uploaded Themes' section is below our Recommended themes.  It was mentioned in #37285 that 'Uploaded Themes' may be best shown above 'Recommended Themes'.  However, the component that renders this "Uploaded Themes" section in the atomic site happens to be the exact same component call used to show all wpcom search results for simple sites (then in the case of atomic the wpcom search results come from an extra child being passed in as props).  

Thus to get this 'Uploaded Themes' section above "Recommended Themes" for atomic sites while hiding the normal wpcom search results for simple sites is one of the changes that may take a bit of restructuring and would be best addressed in a follow up PR if necessary.  I would however argue that below the search bar is a fitting place for these 'Uploaded Themes' since this section is designed to filter based off of the search bar input.
  
Fixes #37285

### CODE REVIEW

File: client/my-sites/themes/theme-showcase.jsx
Comment (noahtallen) [inline client/my-sites/themes/theme-showcase.jsx]: Should this be separated into a new component, and then we show `RecommendedThemes` along with `AllThemes`?

that could also be used if we need to gate the feature to certain users

File: client/my-sites/themes/themes-selection.jsx
Comment (noahtallen) [inline client/my-sites/themes/themes-selection.jsx]: Why do we need to pass both  `recommendedThemes` and `themes` into the component? Is there a way to abstract it so that the component only takes a `themes` prop, and then `recommendedThemes` are given to that prop in the right context? (I may also be confused about what this does :)

File: client/state/themes/actions.js
Comment (noahtallen) [inline client/state/themes/actions.js]: Noice!

File: client/state/themes/reducer.js
Comment (noahtallen) [inline client/state/themes/reducer.js]: I think we usually try to add tests for new parts of the state. You can probably find examples by looking at other reducers :)

**Review by noahtallen — COMMENTED** [review]
😍 This looks great! It works very smoothly for me, with no functionality problems. I do have a couple thoughts, and I hope we get a lot of eyes on this change, since it is a big one and affects a lot of users :)

I imagine the tests are failing because some of them probably test that this themes page looks a certain way, so those should probably be updated to match the new format.


- I notice that when going to "theme info" and then back to the theme list, the recommended themes re-loads. I wonder if we have to run `getRecommendedThemes` every time, since this list does not change very often. Or, we could use the existing themes if they already exist in the state instead of showing the loading indicator -- and then when the query finishes, it can update the state behind the scenes.
- Should we gate this change to a subset of users? It seems like one of the largest changes in our experience issues, so I wonder if we need to contain this change to _only_ people who have FSE-eligible sites. (Note that [this selector can be used](https://github.com/Automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/state/selectors/is-site-using-full-site-editing.js) to detect that with some tweaks. This would work out very will once all of the themes shown here have FSE support.)
- Should we duplicate "recommended themes" in the "all themes" list? It feels weird that the first screen-ful of themes is the same as the ones we were already looking at.

Most of the above, I don't know the answer to, so it'll be good to get more feedback :)

**Comment by apeatling** [comment]
Thanks for your work on this one @Addison-Stavlo! Couple of things:

Screen Shot 2019-11-24 at 8 43 49 AM

cc @shaunandrews for the site title + blurb bit.

**Comment by shaunandrews** [comment]
Maybe something like this:

image

File: client/my-sites/themes/themes-selection.jsx
Comment (Addison-Stavlo) [reply client/my-sites/themes/themes-selection.jsx]: Here we are passing one of the two into the `ThemesList` component as the `themes` prop.  So if we passed `recommendedThemes` into `ThemesSelection` like we do from the `RecommendedThemes` component, then `ThemesList` receives that as the `themes` prop.  If there is no `recommendedThemes` prop (how the component is normally used), then it falls back to pass `props.themes` connection from state.  So in the case of `ThemesList` it is always receiving only one of those as the `themes` prop.   

Or were you saying that the the `ThemesSelection` component itself should be cleaned up in a way that we aren't passing the recommended list down?  I suppose we could make this a condition in the `connect` HOC to determine what `themes` should be so we don't need to have `recommendedThemes` inside the actual component class.  But we do find `recommendedThemes` helpful as a boolean check in the component class to condition things like `incrementPage` function and whether or not we should render certain other components.  So Im not sure if thats going to be very helpful either. 🤔

**Review by Addison-Stavlo — COMMENTED** [review]
@shaunandrews  that looks great.  Is there similar descriptive sentence you would like to add just below "Recommended Themes" in the section above as well?  "These are our latest themes that support the most customization for new users" or something?

**Comment by Addison-Stavlo** [comment]
@apeatling - 
I just updated it to filter the recommended themes out of the search input by default.  I noticed that filtering them out in all cases may cause some un-wanted issues as seen below when people may try to search via the "auto-loading-homepage" tag or search a theme by its specific name:

Screen Shot 2019-11-25 at 3 17 25 PM

 or
 
Screen Shot 2019-11-25 at 3 17 41 PM

To avoid this I have made it so that this will not show recommended themes in the "all themes" section when it is initially opened or reverts back to a blank input/query, but they will show up once the input starts looking for specific results.  I am assuming this/below is the behavior we are looking for?

Screen Shot 2019-11-25 at 3 31 49 PM

Screen Shot 2019-11-25 at 3 32 11 PM

Screen Shot 2019-11-25 at 3 31 18 PM

File: client/my-sites/themes/theme-showcase.jsx
Comment (Addison-Stavlo) [reply client/my-sites/themes/theme-showcase.jsx]: Perhaps it should.  These were all in the ThemeShowcase from the beginning, and not being 100% certain which components may require being hidden or extracted as part of `AllThemes` I haven't taken any steps in that direction yet.  

It's currently also the case that `RecommendedThemes` requires the `AllThemes` part of the code to run to maintain all of its functionality.  This is why I am hiding the `AllThemes` section with `display: none` instead of conditionally loading it in the first place.  When I had it conditioned to not load until the button is pushed, `RecommendedThemes` would have small bugs in its functionality until `AllThemes` was opened and loaded.

I am going to investigate a bit more to try and clean this all up, and in that case creating the `AllThemes` component could be good as well.  At its current state it works for how we want it to work (alongside `AllThemes`), but is lacking in the sense that `Recommended Themes` is not completely stand-alone and would cause issues down the line if we tried to use it somewhere without the `AllThemes` portion.

**Review by Addison-Stavlo — COMMENTED** [review]
@noahtallen 

>I notice that when going to "theme info" and then back to the theme list, the recommended themes re-loads. 

Good catch!  I have updated this and it should no longer trigger any fetch if this is already on state.

> I imagine the tests are failing because some of them probably test that this themes page looks a certain way, so those should probably be updated to match the new format.

Definitely!  I was holding off at first for some feedback so I knew we were on the right direction but I think its safe to update these now.  Will do shortly.

**Comment by apeatling** [comment]
Tested this out, and I think this is mergeable from an experience point of view. One nitpick, remove the colon at the end of the Recommended Themes title.

**Review by apeatling — APPROVED** [review]

**Comment by noahtallen** [comment]
> I think this is mergeable from an experience point of view.

I think it'd be good to write a happiness update about the changes since it is big :)

**Comment by Addison-Stavlo** [comment]
@noahtallen 
> I think it'd be good to write a happiness update about the changes since it is big :)

Definitely a big yes!  I am planning to write a p2 about it before merging for visibility purposes (is a happiness update different or just an update on the happiness p2?).  

@apeatling 
> Tested this out, and I think this is mergeable from an experience point of view. 

Thats great!  I'l update the tests and do a little bit more digging to see if making these components a bit more stand-alone friendly will be a relatively simple fix, but I am glad to know from a experience PoV that this is about ready 😄 

> One nitpick, remove the colon at the end of the Recommended Themes title.

Done.

**Comment by apeatling** [comment]
@Addison-Stavlo We need to add a tracks event on the load more button so we can track this in funnels. /cc @ianstewart

**Comment by Addison-Stavlo** [comment]
It looks like this also effects the logged out themes view such as `WordPress.com/themes` or `calypso.localhost:3000/themes`.  So my first focus today is going to be investigating and making changes as needed.  

In terms of style and functionality for this "Logged Out" themes area, it is working the same as the logged in user section but with its default filter bar above everything (see below).  I am thinking we may want to remove this Recommended Themes section from the logged out view?  @shaunandrews @apeatling how do we feel about this?

Screen Shot 2019-11-27 at 11 39 24 AM

As seen in the screenshot above, the top blue bar applies filters to the themes shown in the "Advanced Themes" section that is hidden behind the button and much further down the page once opened.  So the section of themes that the blue bar filters is so far down the page that you cant see them when you are using the blue bar.

Also, does anybody know of any other areas where a logged out user can view the Theme Showcase?

**Comment by Addison-Stavlo** [comment]
> @Addison-Stavlo We need to add a tracks event on the load more button so we can track this in funnels. /cc @ianstewart

Ok.  Il be happy to add it but I am not totally sure how/what I am adding.  I will look into this more, but do we have any more info or insight on this?

**Comment by apeatling** [comment]
For tracks event, I think you need to call `recordTracksEvent` from https://github.com/Automattic/wp-calypso/tree/master/client/state/analytics when the button is clicked.

The event can be called `calypso_themeshowcase_more_themes_clicked`

File: client/state/themes/reducer.js
Comment (Addison-Stavlo) [reply client/state/themes/reducer.js]: Great suggestion.  I have added some basic tests for the actions and reducers.

**Comment by Addison-Stavlo** [comment]
I was able to update the logged out area not to render the `recommendedThemes` section and to have the old themes search open by default.  So this area now seems to look and function as the live version of `wordpress.com/themes` currently does.

I have also been able to add tests for actions/reducers, get the LoggedOut component tests to pass again, and have added the track event to the "More Themes" button.

**Comment by apeatling** [comment]
This is all looking good to me. :+1:

**Review by vindl — APPROVED** [review]
This worked as described for me. 👍

**Comment by vindl** [comment]
I've looked into some of the test failures below, and it seems like they are related to the changes here: `[WPCOM] Activating Themes...`.

**Review by noahtallen — APPROVED** [review]
It's working well for me! Should be ready to go once the tests are passing :)

**Comment by Addison-Stavlo** [comment]
Hrm, you are right.. while all the standard tests from `npm run test` seem to pass, it looks like there are some e2e tests here failing and they all say `Timed out waiting for element with css selector of 'a[data-e2e-value="free"]' to be clickable`.  I wonder if this has something to do with us hiding the search bar by default...

**Comment by Addison-Stavlo** [comment]
After adding a step in a couple parts of the e2e test to open the button, it looks like there is 1 theme related test not passing now.  It looks like it is coming from here:

https://github.com/Automattic/wp-calypso/blob/d8bf6a036301628802e0853eb771ddfb74e62b3b/test/e2e/specs/wp-theme-multisite-spec.js#L70-L72

**Comment by Addison-Stavlo** [comment]
I did a couple small updates to the e2e tests so that the "more themes" button is clicked where the tests need access to the hidden content.  There is 1 more themes related test not passing and it looks like it is coming from here:
https://github.com/Automattic/wp-calypso/blob/d8bf6a036301628802e0853eb771ddfb74e62b3b/test/e2e/specs/wp-theme-multisite-spec.js#L70-L72

It is timing out waiting for this "Try & Customize" button to appear and be clickable.  I am thinking this accesses the first theme option on the page and tries to do this, but since the first theme option is now a template-first theme (which don't appear to have that option) it times out waiting for a button that will never appear.

I will look into this more tomorrow.  But if that is the case, what are our thoughts about this test?  Is `Try & Customize` something that we are phasing out as we move to FSE?  Is this a section of the themes e2e tests that should be removed?  Or do we work at updating this in some manner that it accesses a non template-first-theme in the bottom section of the screen instead?

**Comment by noahtallen** [comment]
> it times out waiting for a button that will never appear.

Really interesting because the tests did seem to work when we changed that button, and the first theme was a template first theme in that case too. I wonder if clicking "show more" first will make it work? Or if there's a way to target a non-template first theme?

For legacy purposes, it could be good to make the test continue working for legacy themes, but also add a similar test for the template first themes

**Comment by Addison-Stavlo** [comment]
At this point I have fixed most of the e2e tests.  Currently there is 1 failing, saying "Successful activation dialog should show the correct theme in the current theme bar".  The test seems to activate a different theme and then check the Current Theme Bar displays the proper theme.  The test fails and sees the old theme's name present.  However, when I try to do replicate this I am not seeing any issues.  🤔

**Comment by vindl** [comment]
> At this point I have fixed most of the e2e tests. Currently there is 1 failing, saying "Successful activation dialog should show the correct theme in the current theme bar". The test seems to activate a different theme and then check the Current Theme Bar displays the proper theme. The test fails and sees the old theme's name present. However, when I try to do replicate this I am not seeing any issues. 🤔

Looping in @Stojdza for help with e2e tests. :)

**Comment by Stojdza** [comment]
Failing tests activated correct theme (same as before), but it was expecting another theme - a wrong one. This happened because [test was getting the first theme](https://github.com/Automattic/wp-calypso/pull/37696/files?file-filters%5B%5D=.js#diff-6af1f5cc87189fb61b8e9d8310cb7255L65) from recommended Themes list, introduced in this PR. 

I've [updated the selector](https://github.com/Automattic/wp-calypso/pull/37696/commits/7e5af6eeadd966f89606daed262fdc67d0c310da) to be more specific. Now, we are getting the first theme from all themes list, at the bottom of the page.

**Comment by Copons** [comment]
Noting that `ci/circleci: lint-and-translate` seems to fail at lines unrelated to this change.

Thanks @Stojdza for helping us here! I was about to giving it a deeper look, so you spared me what could have been a long investigation 😄 

I've restarted the e2e, finger crossed!

**Comment by Addison-Stavlo** [comment]
@Stojdza - Thank you so much for the help!

> This happened because test was getting the first theme from recommended Themes list, introduced in this PR.

Ok that makes sense now.  I had some other tests failing for a similar reason earlier, I missed that this was the same thing happening here.

> I've updated the selector to be more specific. Now, we are getting the first theme from all themes list, at the bottom of the page.

Thanks again!

**Comment by apeatling** [comment]
🎉Nice work @Addison-Stavlo !

**Comment by vindl** [comment]
Thanks for helping us with e2e test @Stojdza and great work once again @Addison-Stavlo! 🚀

**Pull request merged**
